x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/InternalMultiTerms.java

```diff
@@ -0,0 +1,613 @@
+    }
+
+    @Override
+    protected BucketOrder getOrder() {
+        return order;
+    }
+
+    @Override
+    protected long getSumOfOtherDocCounts() {
+        return otherDocCount;
+    }
+
+    @Override
+    protected long getDocCountError() {
+        return docCountError;
+    }
+
+    @Override
+    protected void setDocCountError(long docCountError) {
+        this.docCountError = docCountError;
+    }
+
+    @Override
+    protected long getMinDocCount() {
+        return minDocCount;
+    }
+
+    @Override
+    protected int getRequiredSize() {
+        return requiredSize;
+    }
+
+    @Override
+    protected Bucket createBucket(long docCount, InternalAggregations aggs, long docCountError, Bucket prototype) {
+        return new Bucket(prototype.terms, docCount, aggs, prototype.showDocCountError, docCountError, formats, keyConverters);
+    }
+
+    @Override
+    public InternalMultiTerms create(List<Bucket> buckets) {
+        return new InternalMultiTerms(
+            name,
+            reduceOrder,
+            order,
+            requiredSize,
+            minDocCount,
+            shardSize,
+            showTermDocCountError,
+            otherDocCount,
+            buckets,
+            docCountError,
+            formats,
+            keyConverters,
+            getMetadata()
+        );
+    }
+
+    /**
+     * Checks if any keys need to be promoted to double from long or unsigned_long
+     */
+    private boolean[] needsPromotionToDouble(List<InternalAggregation> aggregations) {
+        if (aggregations.size() < 2) {
+            return null;
+        }
+        boolean[] promotions = null;
+
+        for (int i = 0; i < keyConverters.size(); i++) {
+            boolean hasLong = false;
+            boolean hasUnsignedLong = false;
+            boolean hasDouble = false;
+            boolean hasNonNumber = false;
+            for (InternalAggregation aggregation : aggregations) {
+                InternalMultiTerms agg = (InternalMultiTerms) aggregation;
+                KeyConverter keyConverter = agg.keyConverters.get(i);
+                switch (keyConverter) {
+                    case DOUBLE:
+                        hasDouble = true;
+                        break;
+                    case LONG:
+                        hasLong = true;
+                        break;
+                    case UNSIGNED_LONG:
+                        hasUnsignedLong = true;
+                        break;
+                    default:
+                        hasNonNumber = true;
+                        break;
+                }
+            }
+            if (hasNonNumber && (hasDouble || hasUnsignedLong || hasLong)) {
+                throw new AggregationExecutionException(
+                    "Merging/Reducing the multi_term aggregations failed when computing the aggregation "
+                        + name
+                        + " because the field in the position "
+                        + (i + 1)
```

Comment:
It might be a script as well... The term "field" here is probably not the best either, but I couldn't make this message concise and clear at the same time.

Reply:
Makes sense